crates/stages/src/stages/index_log_history.rs

```diff
@@ -0,0 +1,352 @@
+use crate::{ExecInput, ExecOutput, Stage, StageError, UnwindInput, UnwindOutput};
+use reth_db::{cursor::DbCursorRO, database::Database, tables, transaction::DbTx, DatabaseError};
+use reth_primitives::{
+    stage::{
+        CheckpointBlockRange, EntitiesCheckpoint, IndexHistoryCheckpoint, StageCheckpoint, StageId,
+    },
+    BlockNumber,
+};
+use reth_provider::DatabaseProviderRW;
+use std::ops::RangeInclusive;
+use tracing::info;
+
+/// The log indexing stage.
+///
+/// This stage walks over all available receipts and inserts the index of block numbers where
+/// particular log addresses and topics occurred.
+///
+/// For more information see [reth_db::tables::LogAddressHistory] and
+/// [reth_db::tables::LogTopicHistory].
+#[derive(Debug, Clone)]
+pub struct IndexLogHistoryStage {
+    commit_threshold: u64,
+}
+
+impl Default for IndexLogHistoryStage {
+    fn default() -> Self {
+        Self { commit_threshold: 100_000 }
+    }
+}
+
+#[async_trait::async_trait]
+impl<DB: Database> Stage<DB> for IndexLogHistoryStage {
+    /// Return the id of the stage
+    fn id(&self) -> StageId {
+        StageId::IndexLogHistory
+    }
+
+    /// Execute the stage.
+    async fn execute(
+        &mut self,
+        provider: &mut DatabaseProviderRW<'_, &DB>,
+        input: ExecInput,
+    ) -> Result<ExecOutput, StageError> {
+        if input.target_reached() {
+            return Ok(ExecOutput::done(input.checkpoint()))
+        }
+
+        let (range, is_final_range) = input.next_block_range_with_threshold(self.commit_threshold);
+
+        let mut stage_checkpoint = stage_checkpoint(provider, input.checkpoint(), &range)?;
+
+        // TODO: stage checkpoint
+        let (log_address_indices, log_topic_indices, num_of_receipts) =
+            provider.get_log_addresses_and_topics(range.clone())?;
+        provider.insert_log_address_history_index(log_address_indices)?;
+        provider.insert_log_topic_history_index(log_topic_indices)?;
```

Comment:
good opportunity to define these as new provider traits IMO vs adding on the struct